GCC 13 compiler warning in kernel/userspace.c

**Describe the bug**
GCC 13 has improved analysis which raises a warning due to reading address outside of `_thread_idx_map`.

If `CONFIG_MAX_THREAD_BYTES` is not a multiple of 4 the cast to `uint32_t*` will read beyond array boundaries.

**To Reproduce**

Steps to reproduce the behavior:

    ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb GNUARMEMB_TOOLCHAIN_PATH=/opt/arm/arm-gnu-toolchain-13.2.Rel1-x86_64-arm-none-eabi ./scripts/twister -T tests/kernel/common -v --hardware-map map.yml  --device-testing -c -s tests/kernel/common/kernel.common

**Expected behavior**

Build without warnings.

**Logs and console output**

```
In file included from /home/david/zephyrproject/zephyr/include/zephyr/arch/arm/arch.h:34,
                 from /home/david/zephyrproject/zephyr/include/zephyr/arch/cpu.h:19,
                 from /home/david/zephyrproject/zephyr/include/zephyr/kernel_includes.h:36,
                 from /home/david/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /home/david/zephyrproject/zephyr/kernel/userspace.c:8:
In function 'sys_clear_bit',
    inlined from 'sys_bitfield_clear_bit' at /home/david/zephyrproject/zephyr/include/zephyr/arch/common/sys_bitops.h:71:2,
    inlined from 'thread_idx_alloc' at /home/david/zephyrproject/zephyr/kernel/userspace.c:280:4,
    inlined from 'z_object_alloc' at /home/david/zephyrproject/zephyr/kernel/userspace.c:401:8:
/home/david/zephyrproject/zephyr/include/zephyr/arch/common/sys_bitops.h:33:18: error: array subscript 'uint32_t {aka volatile unsigned int}[0]' is partly outside array bounds of 'uint8_t[3]' {aka 'unsigned char[3]'} [-Werror=array-bounds=]
   33 |         uint32_t temp = *(volatile uint32_t *)addr;
      |                  ^~~~
/home/david/zephyrproject/zephyr/kernel/userspace.c: In function 'z_object_alloc':
/home/david/zephyrproject/zephyr/kernel/userspace.c:77:16: note: object '_thread_idx_map' of size 3
   77 | extern uint8_t _thread_idx_map[CONFIG_MAX_THREAD_BYTES];
      |                ^~~~~~~~~~~~~~~
In function 'sys_clear_bit',
    inlined from 'sys_bitfield_clear_bit' at /home/david/zephyrproject/zephyr/include/zephyr/arch/common/sys_bitops.h:71:2,
    inlined from 'thread_idx_alloc' at /home/david/zephyrproject/zephyr/kernel/userspace.c:280:4,
    inlined from 'z_object_alloc' at /home/david/zephyrproject/zephyr/kernel/userspace.c:401:8:
/home/david/zephyrproject/zephyr/include/zephyr/arch/common/sys_bitops.h:35:36: error: array subscript 'uint32_t {aka volatile unsigned int}[0]' is partly outside array bounds of 'uint8_t[3]' {aka 'unsigned char[3]'} [-Werror=array-bounds=]
   35 |         *(volatile uint32_t *)addr = temp & ~(1 << bit);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/david/zephyrproject/zephyr/kernel/userspace.c: In function 'z_object_alloc':
/home/david/zephyrproject/zephyr/kernel/userspace.c:77:16: note: object '_thread_idx_map' of size 3
   77 | extern uint8_t _thread_idx_map[CONFIG_MAX_THREAD_BYTES];
      |                ^~~~~~~~~~~~~~~
``

**Environment (please complete the following information):**
 - OS: Ubuntu 22.04
 - Toolchain: gnuarmemb (13.2.Rel1)
 - d77dc62d196b8f5f58622a05414002cbf40285ba

--- a/kernel/userspace.c
+++ b/kernel/userspace.c
@@ -277,8 +277,10 @@ static bool thread_idx_alloc(uintptr_t *tidx)
 		if (idx != 0) {
 			*tidx = base + (idx - 1);
 
-			sys_bitfield_clear_bit((mem_addr_t)_thread_idx_map,
-					       *tidx);
+			/* Clear the bit. We already know the array index,
+			 * and the bit to be cleared.
+			 */
+			_thread_idx_map[i] &= ~(BIT(idx - 1));
 
 			/* Clear permission from all objects */
 			k_object_wordlist_foreach(clear_perms_cb,
@@ -308,7 +310,11 @@ static void thread_idx_free(uintptr_t tidx)
 	/* To prevent leaked permission when index is recycled */
 	k_object_wordlist_foreach(clear_perms_cb, (void *)tidx);
 
-	sys_bitfield_set_bit((mem_addr_t)_thread_idx_map, tidx);
+	/* Figure out which bits to set in _thread_idx_map[] and set it. */
+	int base = tidx / NUM_BITS(_thread_idx_map[0]);
+	int offset = tidx % NUM_BITS(_thread_idx_map[0]);
+
+	_thread_idx_map[base] |= BIT(offset);
 }
 
 static struct k_object *dynamic_object_create(enum k_objects otype, size_t align,

